File: backend/internal/model/waitlist.go
package model

import "time"

// WaitlistEntry 地块候补认养申请实体。
//
// 状态机（WaitlistStatus）：
//
//	waiting   —— 候补排队中（地块已被认养，市民申请候补）
//	invited   —— 地块释放后，最早申请者获得优先认养资格（认养或取消前一直有效）
//	adopted   —— 受邀人在资格有效期内完成认养（终态）
//	cancelled —— 申请人主动取消（终态；同一地块同一人可再次申请）
//
// 同一地块同一用户只保留一条“有效申请”（waiting / invited）：
// service 层在事务内先锁地块行（SELECT ... FOR UPDATE）再查重，
// 另由数据库部分唯一索引 waitlist_active_uniq（plot_id, user_id）
// WHERE status IN ('waiting','invited') 兜底（见 database.EnsurePartialIndexes）。
type WaitlistEntry struct {
	ID         uint       `gorm:"primaryKey" json:"id"`
	PlotID     uint       `gorm:"not null;index:idx_waitlist_plot" json:"plot_id"`
	Plot       *Plot      `gorm:"foreignKey:PlotID" json:"plot"`
	UserID     uint       `gorm:"not null;index:idx_waitlist_user" json:"user_id"`
	User       *User      `gorm:"foreignKey:UserID" json:"user"`
	Status     string     `gorm:"size:32;not null;default:waiting;index:idx_waitlist_status" json:"status"`
	InvitedAt  *time.Time `json:"invited_at"`
	CanceledAt *time.Time `json:"canceled_at"`
	ResolvedAt *time.Time `json:"resolved_at"`
	Note       string     `gorm:"size:512" json:"note"`
	CreatedAt  time.Time  `json:"created_at"`
	UpdatedAt  time.Time  `json:"updated_at"`
}

File: frontend/src/api/waitlist.ts
import { del, get, post } from '@/utils/request'
import type { Plot } from './plot'
import type { UserInfo } from './auth'

export type WaitlistStatusValue = 'waiting' | 'invited' | 'adopted' | 'cancelled'

export interface WaitlistEntry {
  id: number
  plot_id: number
  plot?: Plot | null
  user_id: number
  user?: UserInfo | null
  status: WaitlistStatusValue
  note: string
  rank: number
  invited_at?: string | null
  canceled_at?: string | null
  created_at: string
}

export interface MyWaitlistStatus {
  plot_id: number
  waitlist_id: number
  status: WaitlistStatusValue | ''
  rank: number
  waiting_count: number
  active_count: number
  reserved: boolean
  invited_to_me: boolean
}

export interface WaitlistSummary extends MyWaitlistStatus {
  mine: WaitlistEntry | null
}

// 申请候补
export function applyWaitlist(plotId: number, note = ''): Promise<WaitlistEntry> {
  return post(`/plots/${plotId}/waitlist`, { note })
}

// 取消自己的候补申请
export function cancelWaitlist(id: number): Promise<WaitlistEntry> {
  return del(`/waitlist/${id}`)
}

// 管理员/认养人查看候补名单（按申请时间排序）
export function listPlotWaitlist(plotId: number): Promise<{ list: WaitlistEntry[]; total: number }> {
  return get(`/plots/${plotId}/waitlist`)
}

// 候补人数 + 当前用户申请状态
export function getWaitlistSummary(plotId: number): Promise<WaitlistSummary> {
  return get(`/plots/${plotId}/waitlist/summary`)
}

// 地块列表批量查询
export function batchWaitlistStatus(plotIds: number[]): Promise<{ list: MyWaitlistStatus[] }> {
  return get('/waitlist/status', { params: { plot_ids: plotIds.join(',') } })
}

// 我的候补申请
export function listMyWaitlist(params?: Record<string, any>): Promise<{ list: WaitlistEntry[]; total: number; page: number; page_size: number }> {
  return get('/waitlist/mine', { params })
}
